### PR TITLE
skipping: refactor mark evaluation

### PR DESCRIPTION
**Pull request by bluetech**
Recently I had a look at the skipping plugin and had a very hard time following the code, so I decided to clean it up.

The PR is split into 3 commits, it should be easier to review each on its own (and see the commit messages).

### CODE REVIEW

**Comment by bluetech** [comment]
The coverage is a bit lacking, I'll add tests to cover the missing bits later.

File: src/_pytest/skipping.py [hunk 1]
```diff
@@ -62,81 +67,210 @@ def nop(*args, **kwargs):
...
+
+    Returns (result, reason). The reason is only relevant if the result is True.
+    """
+    # String condition.
+    if isinstance(condition, str):
+        globals_ = {
+            "os": os,
+            "sys": sys,
+            "platform": platform,
+            "config": item.config,
```
Comment (RonnyPfannschmidt) [inline src/_pytest/skipping.py:102]: followup idea - lets put `item` in as feature

File: src/_pytest/skipping.py [hunk 2]
```diff
@@ -62,81 +67,210 @@ def nop(*args, **kwargs):
...
+                "Error evaluating %r: " % mark.name
+                + "you need to specify reason=STRING when using booleans as conditions."
+            )
+            fail(msg, pytrace=False)
+
+        try:
+            result = bool(condition)
+        except Exception as exc:
+            msglines = [
+                "Error evaluating %r condition as a boolean" % mark.name,
```
Comment (RonnyPfannschmidt) [inline src/_pytest/skipping.py:139]: followup note: safe_repr of the condition if possible?

File: src/_pytest/skipping.py [hunk 3]
```diff
@@ -62,81 +67,210 @@ def nop(*args, **kwargs):
...
+        globals_ = {
+            "os": os,
+            "sys": sys,
+            "platform": platform,
+            "config": item.config,
+        }
+        if hasattr(item, "obj"):
+            globals_.update(item.obj.__globals__)  # type: ignore[attr-defined]
+        try:
+            condition_code = _pytest._code.compile(condition, mode="eval")
```
Comment (RonnyPfannschmidt) [inline src/_pytest/skipping.py:107]: another followup idea, should condition as `lambda : ... or lambda item: ..` be enabled

**Comment by bluetech** [comment]
I added tests which should hopefully close the coverage gaps.

While doing this, I also noticed another thing to simplify -- please see the new commit. It also ends up improving performance a bit.

@RonnyPfannschmidt regarding you follow up ideas, I think I see where you're going with that. I'm going to open an issue to discuss.

File: src/_pytest/skipping.py (hunk 2 above)
Comment (bluetech) [reply src/_pytest/skipping.py:131]: Usually it will just be `True`/`False` (and if not, it *should* be), so I don't think it's worth it.

**Comment by bluetech** [comment]
In that last commit I removed the hookwrapper, but that was wrong, the NOTRUN check must be guaranteed to run before the actual call, of course, so I kept it as a hookwrapper.

Also applied @nicoddemus suggestion to update the links.

**Comment by sallner** [comment]
That is much clearer, but it would have been nice to have a changelog entry. It was some search to find the pytest version which supports it. This was needed for https://github.com/pytest-dev/pytest-rerunfailures/pull/162.

**Comment by RonnyPfannschmidt** [comment]
@sallner i can see where you are coming from, but this is also a function we explicitly did not expose yet

**Pull request merged**